### PR TITLE
Fix race condition where Encapsulate Field could silently fail

### LINKED ISSUE #19816: CSharpEncapsulateField.EncapsulateThroughCommand integration test failure
https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_vs-integration_prtest/4673/testReport/junit/Roslyn.VisualStudio.IntegrationTests.CSharp/CSharpEncapsulateField/EncapsulateThroughCommand/

```
MESSAGE:
Assert.Contains() Failure\r\nNot found: public static int? Param { get => param; set => param = value; }\r\nIn value: \r\nnamespace myNamespace\r\n{\r\n class Program\r\n {\r\n private static int? param = 0;\r\n static void Main(string[] args)\r\n {\r\n param = 80;\r\n }\r\n }\r\n}
+++++++++++++++++++
STACK TRACE:
at Microsoft.VisualStudio.IntegrationTest.Utilities.OutOfProcess.Editor_OutOfProc.Verifier.TextContains(String expectedText, Boolean assertCaretPosition) at Roslyn.VisualStudio.IntegrationTests.CSharp.CSharpEncapsulateField.EncapsulateThroughCommand() in q:\roslyn\src\VisualStudio\IntegrationTest\IntegrationTests\CSharp\CSharpEncapsulateField.cs:line 47
```

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #19816

<details><summary>Ask Mode template not completed</summary>

### Customer scenario

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

### Bugs this fixes

(either VSO or GitHub links)

### Workarounds, if any

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

### Risk

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

### Performance impact

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

### Root cause analysis

How did we miss it?  What tests are we adding to guard against it in the future?

### How was the bug found?

(E.g. customer reported it vs. ad hoc testing)

### Test documentation updated?

If this is a new non-compiler feature or a significant improvement to an existing feature, update https://github.com/dotnet/roslyn/wiki/Manual-Testing once you know which release it is targeting.

</details>

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs [hunk 1]
```diff
@@ -83,22 +82,30 @@ private bool Execute(EncapsulateFieldCommandArgs args, IUIThreadOperationScope w
 
                 var service = document.GetLanguageService<AbstractEncapsulateFieldService>();
```
Comment (sharwell) [inline src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs]: :bulb: The following code can be added here to increase the failure rate on local developer machines (which tend to be faster than the CI machines used by Jenkins:

```csharp
async Task<int> AsyncOperation(CancellationToken ct)
{
    await Task.Delay(TimeSpan.FromSeconds(2), ct).ConfigureAwait(false);
    return 3;
}

AsyncOperation(cancellationToken).WaitAndGetResult(cancellationToken);
```

File: src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs [hunk 2]
```diff
@@ -83,22 +82,30 @@ private bool Execute(EncapsulateFieldCommandArgs args, IUIThreadOperationScope w
...
 
-                var result = service.EncapsulateFieldAsync(document, spans.First().Span.ToTextSpan(), true, cancellationToken).WaitAndGetResult(cancellationToken);
+                // This is the last point where the operation can be canceled by the operation context
+                // managed by the command system. Make sure to not ignore a cancellation request which
+                // occurred prior to this line before proceeding with the Roslyn-managed dialogs.
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs]: why do we need a roslyn-managed dialog?

File: src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs [hunk 3]
```diff
@@ -83,22 +82,30 @@ private bool Execute(EncapsulateFieldCommandArgs args, IUIThreadOperationScope w
...
+                cancellationToken.ThrowIfCancellationRequested();
 
                 // We are about to show a modal UI dialog so we should take over the command execution
                 // wait context. That means the command system won't attempt to show its own wait dialog 
                 // and also will take it into consideration when measuring command handling duration.
                 waitScope.Context.TakeOwnership();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs]: Woudln't it make more sense to leave it as it was, but only do the .TakeOwnership right before we show the preview dialog?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs]: Note: if we do that, we would still want to do this work before showing hte preview:

```
result.GetNameAsync(cancellationToken).WaitAndGetResult(cancellationToken),
result.GetGlyphAsync(cancellationToken).WaitAndGetResult(cancellationToken),
```

File: src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs [hunk 4]
```diff
@@ -84,23 +83,30 @@ private bool Execute(EncapsulateFieldCommandArgs args, IUIThreadOperationScope w
...
 
+                // The cancellation token associated with the operation scope may or may not have been
+                // canceled when ownership of the context was taken. Treat the token as invalid and use
+                // a different one for future operations in this command handler.
+                cancellationToken = CancellationToken.None;
+
                 if (result == null)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs]: consider extracting the rest of this into a function which deliberately does not take a cancellation token.  doc the function that that's intentional.

File: src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs [hunk 5]
```diff
@@ -84,23 +83,30 @@ private bool Execute(EncapsulateFieldCommandArgs args, IUIThreadOperationScope w
...
 
                 // We are about to show a modal UI dialog so we should take over the command execution
                 // wait context. That means the command system won't attempt to show its own wait dialog 
                 // and also will take it into consideration when measuring command handling duration.
                 waitScope.Context.TakeOwnership();
 
+                // The cancellation token associated with the operation scope may or may not have been
+                // canceled when ownership of the context was taken. Treat the token as invalid and use
+                // a different one for future operations in this command handler.
+                cancellationToken = CancellationToken.None;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs]: this is still a good idea.  just so we don't have later code taht accidently uses the token.  however, if we extract a helper function, we can make it even harder to do that.

File: src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs]: agreed.  i would imagine we have this problem in a few other places.

**Comment by sharwell** [comment]
Marking as blocked; issue is not fixed

File: src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs [hunk 6]
```diff
@@ -84,23 +83,30 @@ private bool Execute(EncapsulateFieldCommandArgs args, IUIThreadOperationScope w
...
                     notificationService.SendNotification(EditorFeaturesResources.Please_select_the_definition_of_the_field_to_encapsulate, severity: NotificationSeverity.Error);
                     return false;
                 }
 
-                waitScope.AllowCancellation = false;
-
-                var finalSolution = result.GetSolutionAsync(cancellationToken).WaitAndGetResult(cancellationToken);
-
                 var previewService = workspace.Services.GetService<IPreviewDialogService>();
                 if (previewService != null)
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs]: [pretend this comment is on line 117]

The calls there are still using the cancellationToken, which now "looks" like a bug. Your reassigning of it means it's _not_ a bug, but this still looks very smelly/incomplete. I'm with @CyrusNajmabadi's suggestion to just move that work up, then this "appears" more complete without a careful analysis.

**Comment by sharwell** [comment]
Superseded by #61531

**Pull request closed without merging**